src/dtls.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2020 Cisco and/or its affiliates.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+/*
+ * This is a minimal DTLS implementation intended only to parse the server name
+ * extension.  This was created based primarily on Wireshark dissection of a
+ * DTLS handshake and RFC4366.
+ */
+
+#include <stdio.h>
+#include <stdlib.h> /* malloc() */
+#include <stdint.h>
+#include <string.h> /* strncpy() */
+#include <sys/socket.h>
+#include <sys/types.h>
+#include "dtls.h"
+#include "sni.h"
+#include "protocol.h"
+#include "logger.h"
+
+#define DTLS_HANDSHAKE_CONTENT_TYPE      0x16
+#define DTLS_VERSION_12                  0xfefd
+#define DTLS_HEADER_LEN                  3
+#define DTLS_HANDSHAKE_TYPE_CLIENT_HELLO 0x01
+
+#ifndef MIN
+#define MIN(X, Y) ((X) < (Y) ? (X) : (Y))
+#endif
+
+static int parse_dtls_header(const uint8_t*, size_t, char **);
+
+static const char dtls_alert[] = {
+    0x15, /* DTLS Alert */
+    0x02, 0x01, /* DTLS version  */
+    0x00, 0x02, /* Payload length */
+    0x02, 0x28, /* Fatal, handshake failure */
+};
+
+const struct Protocol *const dtls_protocol = &(struct Protocol){
+    .name = "dtls",
+    .default_port = 443,
+    .parse_packet = (int (*const)(const char *, size_t, char **))&parse_dtls_header,
+    .abort_message = dtls_alert,
+    .abort_message_len = sizeof(dtls_alert)
+};
+
+
+/* Parse a DTLS packet for the Server Name Indication extension in the client
+ * hello handshake, returning the first servername found (pointer to static
+ * array)
```

Comment:
`parse_server_name_extension` allocates a string with `malloc()` and returns it via the `hostname` argument which must be freed by the caller, not a static array.